### PR TITLE
statistics: every histogram use its own version to update.

### PR DESCRIPTION
**Pull request by hanfei1991**
after we support delta updae in #3053 , we shouldn't load the whole histogram when we find the table version updated. Every histogram has it's own version.
@shenli @coocood @zimulala @lamxTyler  PTAL

### CODE REVIEW

**Comment by lamxTyler** [comment]
LGTM

File: statistics/table.go [hunk 1]
```diff
@@ -49,60 +48,84 @@ type Table struct {
...
+		nt.Indices[id] = idx
+	}
+	return nt
+}
+
 // SaveToStorage saves stats table to storage.
-func (h *Handle) SaveToStorage(ctx context.Context, t *Table) error {
-	_, err := ctx.(sqlexec.SQLExecutor).Execute("begin")
+func (h *Handle) SaveToStorage(t *Table) error {
+	_, err := h.ctx.(sqlexec.SQLExecutor).Execute("begin")
```
Comment (coocood) [inline statistics/table.go]: define `exec := h.ctx.(sqlexec.SQLExecutor)` can simplify the code

File: statistics/table.go [hunk 2]
```diff
@@ -113,51 +136,56 @@ func (h *Handle) TableStatsFromStorage(ctx context.Context, info *model.TableInf
...
 				}
 			}
 			if idx != nil {
-				table.Indices[idx.ID] = idx
+				table.Indices[histID] = idx
 				indexCount++
 			} else {
-				log.Warnf("We cannot find index id %d in table %s now. It may be deleted.", histID, info.Name)
+				// TODO: If we support create index ddl, it may be a new created index. We needn't refer tableInfo any more.
+				log.Warnf("We cannot find index id %d in table %s now. It may be deleted.", histID, tableInfo.Name)
```
Comment (coocood) [inline statistics/table.go]: When we adding a new index, this always happen, do we really need to log at warn level?

File: statistics/table.go [hunk 3]
```diff
@@ -113,51 +136,56 @@ func (h *Handle) TableStatsFromStorage(ctx context.Context, info *model.TableInf
...
-					col = &Column{Histogram: *hg}
 					break
 				}
 			}
 			if col != nil {
 				table.Columns[col.ID] = col
 				columnCount++
 			} else {
-				log.Warnf("We cannot find column id %d in table %s now. It may be deleted.", histID, info.Name)
+				log.Warnf("We cannot find column id %d in table %s now. It may be deleted.", histID, tableInfo.Name)
```
Comment (coocood) [inline statistics/table.go]: ditto

File: statistics/table.go (hunk 2 above)
Comment (hanfei1991) [inline statistics/table.go]: When we add a new index, the stats table will be updated later.

File: statistics/table.go (hunk 3 above)
Comment (hanfei1991) [inline statistics/table.go]: see #3061 This will not happend too much,

**Comment by hanfei1991** [comment]
@coocood I have added some tests. PTAL

File: statistics/table.go [hunk 4]
```diff
@@ -113,51 +136,56 @@ func (h *Handle) TableStatsFromStorage(ctx context.Context, info *model.TableInf
 	for _, row := range rows {
 		distinct := row.Data[3].GetInt64()
 		histID := row.Data[2].GetInt64()
+		histVer := row.Data[4].GetUint64()
 		if row.Data[1].GetInt64() > 0 {
 			// process index
-			var idx *Index
-			for _, idxInfo := range info.Indices {
+			idx := table.Indices[histID]
+			for _, idxInfo := range tableInfo.Indices {
```
Comment (tiancaiamao) [inline statistics/table.go]: I think you can extract a function that getIndexByID here.

File: statistics/table.go [hunk 5]
```diff
@@ -113,51 +136,56 @@ func (h *Handle) TableStatsFromStorage(ctx context.Context, info *model.TableInf
...
-				log.Warnf("We cannot find index id %d in table %s now. It may be deleted.", histID, info.Name)
+				// TODO: If we support create index ddl, it may be a new created index. We needn't refer tableInfo any more.
+				log.Warnf("We cannot find index id %d in table %s now. It may be deleted.", histID, tableInfo.Name)
 			}
 		} else {
 			// process column
-			var col *Column
-			for _, colInfo := range info.Columns {
+			col := table.Columns[histID]
+			for _, colInfo := range tableInfo.Columns {
```
Comment (tiancaiamao) [inline statistics/table.go]: getColumnByID ?

File: statistics/table.go (hunk 4 above)
Comment (hanfei1991) [inline statistics/table.go]: tableInfo will not be used when load stats in future. this function will not benifit a lot.

Comment (hanfei1991) [inline statistics/table.go]: This part of code will refactor after we can process the psuedo column correctly.

File: statistics/table.go (hunk 5 above)
Comment (hanfei1991) [inline statistics/table.go]: ditto

**Comment by coocood** [comment]
@hanfei1991 
There is a cheaper way to verify the histogram is not reloaded.
record the pointer of the histogram before update, 
assert the pointer didn't change after update.

**Comment by hanfei1991** [comment]
@coocood Your method makes code more complex. We can log Load***Count in future, it's a more intuitive way.

**Comment by hanfei1991** [comment]
@coocood PTAL

**Comment by coocood** [comment]
LGTM

**Pull request merged**